RenameIndependentSubregs: Add missing sub-range for new IMPLICIT_DEFs

petar-avramovic opened the pull request:
Existing sub-ranges are correctly updated because new IMPLICIT_DEF is
added, but there is missing sub-range for IMPLICIT_DEF itself.
Because of missing sub-range in live-intervals for IMPLICIT_DEF,
register allocator does not know that IMPLICIT_DEF rewrites its
virtual sub-registers and can end up assigning overlapping physical
registers to them.
This results in deleting instructions that were defined by sub-registers
overwritten by IMPLICIT_DEF as they are now dead.

arsenm reviewed (commented):
Needs a new verifier mir test

jayfoad commented:
What was the effect of the bug? Was it failing MachineVerifier, or just generating incorrect code? Do you have a small test case? Or do the tests updated in this PR show the fix - and if so, can you point out where?

petar-avramovic commented:
> What was the effect of the bug? Was it failing MachineVerifier, or just generating incorrect code? Do you have a small test case? Or do the tests updated in this PR show the fix - and if so, can you point out where?

Incorrect code for GlobalISel. It was deleting copies from sgpr to vgpr, sdag was moving everything to VALU so it was not affected by this bug. Instruction deletion shows on some very large tests, change is visible in live intervals if you dump them. Will look into writing mir test.

petar-avramovic commented on llvm/test/CodeGen/AMDGPU/live-interval-bug-in-rename-independent-subregs.mir:
```diff
@@ -4054,22 +4054,22 @@ body:             |
   ; GCN-NEXT:   liveins: $sgpr4, $sgpr5, $sgpr6, $sgpr7, $sgpr33, $vgpr0, $sgpr8_sgpr9:0x000000000000000F, $sgpr10_sgpr11, $sgpr12_sgpr13, $sgpr0_sgpr1_sgpr2_sgpr3
   ; GCN-NEXT: {{  $}}
   ; GCN-NEXT:   renamable $sgpr14 = S_OR_B32 renamable $sgpr33, 2, implicit-def dead $scc
-  ; GCN-NEXT:   renamable $vgpr3 = COPY killed renamable $sgpr14
```
this $vgpr3

petar-avramovic commented on llvm/test/CodeGen/AMDGPU/live-interval-bug-in-rename-independent-subregs.mir:
```diff
@@ -4054,22 +4054,22 @@ body:             |
-  ; GCN-NEXT:   renamable $vgpr3 = COPY killed renamable $sgpr14
-  ; GCN-NEXT:   renamable $vgpr8_vgpr9 = IMPLICIT_DEF
-  ; GCN-NEXT:   renamable $vgpr1_vgpr2 = IMPLICIT_DEF
-  ; GCN-NEXT:   renamable $vgpr0_vgpr1_vgpr2_vgpr3_vgpr4_vgpr5_vgpr6_vgpr7 = IMPLICIT_DEF
```
is overwritten by implicit def here

petar-avramovic commented on llvm/test/CodeGen/AMDGPU/live-interval-bug-in-rename-independent-subregs.mir:
```diff
@@ -4054,22 +4054,22 @@ body:             |
+  ; GCN-NEXT:   renamable $vgpr15 = COPY killed renamable $sgpr14
+  ; GCN-NEXT:   renamable $vgpr9_vgpr10 = IMPLICIT_DEF
+  ; GCN-NEXT:   renamable $vgpr13_vgpr14 = IMPLICIT_DEF
+  ; GCN-NEXT:   renamable $vgpr1_vgpr2_vgpr3_vgpr4_vgpr5_vgpr6_vgpr7_vgpr8 = IMPLICIT_DEF
```
copy is in $vgpr15, no overwrite

arsenm commented on llvm/test/CodeGen/AMDGPU/live-interval-bug-in-rename-independent-subregs.mir:
```diff
@@ -0,0 +1,5426 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
+# RUN: llc -mtriple=amdgcn--amdpal -mcpu=gfx1010 -start-before=rename-independent-subregs -stop-before=amdgpu-mark-last-scratch-load %s -o - | FileCheck -check-prefix=GCN %s
+
+--- |
```
This test needs to be reduced, preferably to something which doesn't require an IR or registers section and at most a handful of blocks

petar-avramovic commented on llvm/test/CodeGen/AMDGPU/live-interval-bug-in-rename-independent-subregs.mir:
```diff
@@ -245,40 +245,41 @@ body: |
-  ; DeadInstDelete-NEXT:   renamable $vgpr8_vgpr9 = IMPLICIT_DEF
-  ; DeadInstDelete-NEXT:   renamable $vgpr1_vgpr2 = IMPLICIT_DEF
-  ; DeadInstDelete-NEXT:   renamable $vgpr0_vgpr1_vgpr2_vgpr3_vgpr4_vgpr5_vgpr6_vgpr7 = IMPLICIT_DEF
+  ; DeadInstDelete-NEXT:   renamable $vgpr15 = COPY killed renamable $sgpr15
```
Added run line that shows instruction that was deleted.
Here, after the change, it is not deleted. 
For context this is lowered divergent phi with uniform inputs, inputs stayed in sgpr but copy to vgpr(phi itself) got deleted because it became dead after register allocator that did not have correct live intervals information.

arsenm commented on llvm/test/CodeGen/AMDGPU/live-interval-bug-in-rename-independent-subregs.mir:
```diff
@@ -0,0 +1,475 @@
+    successors: %bb.82(0x40000000), %bb.40(0x40000000)
+    liveins: $sgpr0, $sgpr1, $sgpr2, $sgpr3, $sgpr4, $sgpr5, $sgpr6, $sgpr7, $vgpr0, $vgpr1, $vgpr2
+
+    %11:vgpr_32 = COPY $vgpr0
```
Can you shrink this down any more? Also, use run-pass=none to compact the register numbers

arsenm commented on llvm/test/CodeGen/AMDGPU/live-interval-bug-in-rename-independent-subregs.mir:
```diff
@@ -0,0 +1,475 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
+# RUN: llc -mtriple=amdgcn--amdpal -mcpu=gfx1010 -start-before=rename-independent-subregs -mattr=-wavefrontsize32,+wavefrontsize64 -stop-before=amdgpu-mark-last-scratch-load %s -o - | FileCheck -check-prefix=RegAlloc %s
+# RUN: llc -mtriple=amdgcn--amdpal -mcpu=gfx1010 -start-before=rename-independent-subregs -mattr=-wavefrontsize32,+wavefrontsize64 -stop-after=machine-cp %s -o - | FileCheck -check-prefix=DeadInstDelete %s
```
You never need -wavefrontsize32

arsenm replied:
Check prefixes should also be all caps

petar-avramovic committed:
RenameIndependentSubregs: Add missing sub-range for new IMPLICIT_DEFs

Existing sub-ranges are correctly updated because new IMPLICIT_DEF is
added, but there is missing sub-range for IMPLICIT_DEF itself.
Because of missing sub-range in live-intervals for IMPLICIT_DEF,
register allocator does not know that IMPLICIT_DEF rewrites its
virtual sub-registers and can end up assigning overlapping physical
registers to them.
This results in deleting instructions that were defined by sub-registers
overwritten by IMPLICIT_DEF as they are now dead.

arsenm reviewed (approved):
Thanks, this is a lot shorter
